concrete/jobs/generate_sitemap.php

```diff
@@ -1,243 +1,78 @@
-     * @param int $cID The page collection id.
-     * @param array $instances An array with some already instantiated helpers, models, ...
+     * @throws \Exception throws an exception in case of errors
      *
-     * @throws \Exception Throws an exception in case of errors.
+     * @return string returns a string describing the job result in case of success
      */
-    private static function addPage($xmlDoc, $cID, $instances)
+    public function run()
     {
-        $page = Page::getByID($cID, 'ACTIVE');
-        if (!static::canIncludePageInSitemap($page, $instances)) {
-            return;
-        }
-        $lastmod = new DateTime($page->getCollectionDateLastModified());
-        $changefreq = $page->getAttribute($instances['ak_sitemap_changefreq']);
-        $priority = $page->getAttribute($instances['ak_sitemap_priority']);
-        $xmlNode = $xmlDoc->addChild('url');
-        $xmlNode->addChild(
-            'loc',
-            $instances['navigation']->getLinkToCollection($page)
-        );
-        $xmlNode->addChild('lastmod', $lastmod->format(DateTime::ATOM));
-        $xmlNode->addChild(
-            'changefreq',
-            Core::make('helper/validation/strings')->notempty($changefreq) ? $changefreq : Config::get('concrete.sitemap_xml.frequency')
-        );
-        $xmlNode->addChild(
-            'priority',
-            is_numeric($priority) ? $priority : Config::get('concrete.sitemap_xml.priority')
-        );
-        if ($instances['multilingualSections']) {
-            $thisSection = MultilingualSection::getBySectionOfSite($page);
-            if (is_object($thisSection) && !$thisSection->isError()) {
-                foreach ($instances['multilingualSections'] as $section) {
-                    $relatedPageID = $section->getTranslatedPageID($page);
-                    if ($relatedPageID) {
-                        $relatedPage = Page::getByID($relatedPageID);
-                        if (static::canIncludePageInSitemap($relatedPage, $instances)) {
-                            $xmlAltNode = $xmlNode->addChild('link', null, 'http://www.w3.org/1999/xhtml');
-                            $xmlAltNode->addAttribute('rel', 'alternate');
-                            $xmlAltNode->addAttribute('hreflang', strtolower(str_replace('_', '-', $section->getLocale())));
-                            $xmlAltNode->addAttribute('href', $instances['navigation']->getLinkToCollection($relatedPage));
-                        }
-                    }
-                }
+        $writer = $this->app->make(SitemapWriter::class);
+        $generator = $writer->getSitemapGenerator();
+        if ($generator->getSiteCanonicalUrl() === '') {
+            $request = $this->app->make(Request::class);
+            if (!$request->getHost()) {
+                throw new UserMessageException(t('Canonical URL is not set and no HTTP request to retrieve the site URL from.'));
```

Comment:
'Canonical URL is not set and **there is** no HTTP request to retrieve the site URL from.'?